File: java/secret-handshake/src/main/java/HandshakeCalculator.java
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class HandshakeCalculator {
    List<Signal> calculateHandshake(final int number) {
        return Stream.of(Signal.values())
                .filter(signal -> signal.isSignalSet(number))
                .sorted(getSignalSortOrder(number))
                .collect(Collectors.toList());
    }

    private static Comparator<? super Signal> getSignalSortOrder(final int number) {
        return Action.REVERSE.isActionSet(number)
                ? Comparator.reverseOrder()
                : Comparator.naturalOrder();
    }
}

File: kotlin/secret-handshake/src/main/kotlin/HandshakeCalculator.kt
package HandshakeCalculator

import Signal

fun calculateHandshake(number: Int): List<Signal> = Signal.values()
        .filter { signal -> signal.isSet(number) }
        .sortedBy { signal -> signal.value * if (reverseSignals(number)) -1 else 1 }

private fun reverseSignals(number: Int): Boolean = (number and 16) == 16
